### PR TITLE
WIP: chore(actions): support cron schedule task

### PR DESCRIPTION
**Pull request by appleboy**
1. only support the default branch in the repository setting.
2. autoload schedule data from the schedule table after starting the service.
3. support specific syntax like `@yearly`, `@monthly`, `@weekly`, `@daily`, `@hourly`

## How to use

See the [GitHub Actions document](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) for getting more detailed information.

```yaml
on:
  schedule:
    - cron: '30 5 * * 1,3'
    - cron: '30 5 * * 2,4'

jobs:
  test_schedule:
    runs-on: ubuntu-latest
    steps:
      - name: Not on Monday or Wednesday
        if: github.event.schedule != '30 5 * * 1,3'
        run: echo "This step will be skipped on Monday and Wednesday"
      - name: Every time
        run: echo "This step will always run"
```

Signed-off-by: Bo-Yi.Wu <appleboy.tw@gmail.com>

### CODE REVIEW

File: models/actions/schedule.go
Comment (wolfogre) [inline models/actions/schedule.go]: Maybe there should be some comments to explain that it's https://github.com/robfig/cron/blob/bc59245fe10efaed9d51b56900192527ed733435/cron.go#L53 and why we have to store it.

File: models/actions/schedule.go
Comment (wolfogre) [inline models/actions/schedule.go]: According to the [doc](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule):
- Event should be always `schedule`.
- CommitSHA should be the last commit on default branch.
- Ref should be always the default branch.

So they should not be stored in DB.

And I didn't check other fields, maybe I'm missing something.

File: models/actions/schedule.go
Comment (wolfogre) [inline models/actions/schedule.go]: The `id` doesn't seem to be used.

File: models/actions/schedule.go
Comment (wolfogre) [inline models/actions/schedule.go]: Maybe:

```go
db.GetEngine(ctx).Insert(rows...)
```

File: services/actions/notifier_helper.go
Comment (wolfogre) [inline services/actions/notifier_helper.go]: Could you please add the new functions to the end of this file or a new file? I would appreciate it.

If you see the diff page, you can find that it's very difficult to review.

Comment (wolfogre) [reply services/actions/notifier_helper.go]: I noticed that there's a better place to go: `schedule.go`

File: services/actions/schedule.go
Comment (wolfogre) [inline services/actions/schedule.go]: After reading the [code](https://github.com/robfig/cron/blob/bc59245fe10efaed9d51b56900192527ed733435/cron.go#L158), I understand why we should do this, there's no way to specify the id of the new entry.

But I still think it's not a good idea to reset all schedules in DB.

It doesn't block, maybe we can improve it in the future.

File: services/actions/schedule.go
Comment (wolfogre) [inline services/actions/schedule.go]: If there's something like:
```
on:
  schedule:
    - cron:  '*/2 * * * *'
    - cron:  '*/3 * * * *'
```

When it's `00:06:00`, so the task will be created twice, right?

That's weird to me, I don't know what GitHub does, maybe we should check it.

**Comment by wolfogre** [comment]
I have an alternative plan for reference:

Define the tables like:
```go
type ActionSchedule struct {
	ID            int64
	Title         string
	Specs         []string
	// and other necessary fields, without EntryIDs
}

type ActionScheduleSpec struct {
	ID         int64
	ScheduleID int64
	Spec       string
}
```

And get rid of `robfig/cron`, just use the current cron frame work of Gitea. We need only one cronjob:
- It runs once per minute.
- Load all rows of `ActionScheduleSpec` from DB, I think it's fine when there are only 3 fields. Or we can have a cache for them.
- Check `Spec` of `ActionScheduleSpec`, record the`ScheduleID` if the spec matches the present time.
- Load `ActionSchedule` from DB by the IDs recorded.
- Create tasks for them.

So we don't have to do something like `resetSchedule` or `schedule.Remove(row.EntryIDs)`.

File: services/actions/schedule.go
Comment (appleboy) [reply services/actions/schedule.go]: Yes. it will occur twice at the same time.

**Comment by zeripath** [comment]
I don't understand why you're using robfig's cron instead of using the inbuilt cron within gitea itself?

**Comment by appleboy** [comment]
@zeripath I will remove the robfig's cron package and refactor the PR, get back to you asap.

File: models/actions/schedule.go
Comment (appleboy) [reply models/actions/schedule.go]: remove EntryIDs and create new table `action_schedule_spec`

File: models/actions/schedule.go
Comment (lunny) [reply models/actions/schedule.go]: But they should still be stored because main branch maybe changed and branch maybe move to a new commit.

File: models/actions/schedule.go
Comment (lunny) [inline models/actions/schedule.go]: What's the owner of a schedule?

File: models/actions/schedule.go
Comment (lunny) [inline models/actions/schedule.go]: If len(rows) <= 0, then just return.

File: models/actions/schedule_spec.go
Comment (lunny) [inline models/actions/schedule_spec.go]: Index ?

**Comment by appleboy** [comment]
@zeripath @wolfogre Done. Please help to review

File: services/actions/notifier_helper.go
Comment (lunny) [inline services/actions/notifier_helper.go]: Why remove this function?

File: services/actions/schedule_tasks.go
Comment (lunny) [inline services/actions/schedule_tasks.go]: For a big instance, how many the records do you think it will contain?

File: services/actions/notifier_helper.go
Comment (appleboy) [reply services/actions/notifier_helper.go]: I will roll back the block.

File: services/actions/notifier_helper.go
Comment (wolfogre) [inline services/actions/notifier_helper.go]: Is it possible to move `CreateScheduleTask` to `schedule_tasks.go`? I don't think it has anything to do with `nofifier_helper.go`, and it makes this file very difficult to review. 😂

File: services/actions/schedule_tasks.go
Comment (wolfogre) [reply services/actions/schedule_tasks.go]: Any better ideas? I think we have to fetch all specs and check them.

File: services/cron/tasks_actions.go
Comment (wolfogre) [inline services/cron/tasks_actions.go]: I don't think we need `RunAtStart` for cronjobs.
```suggestion
		RunAtStart: false,
```

File: services/cron/tasks_actions.go
Comment (wolfogre) [inline services/cron/tasks_actions.go]: BTW, why not add this function to the end of this file when you calling it at the end of `initActionsTasks()`?

File: services/actions/schedule_tasks.go
Comment (wolfogre) [inline services/actions/schedule_tasks.go]: I think it's redundant.
```suggestion
```

File: services/actions/schedule_tasks.go
Comment (wolfogre) [inline services/actions/schedule_tasks.go]: Nop, should use `cron.ParseStandard` here, `cron.Parse` uses

```go
var defaultParser = NewParser(
	Second | Minute | Hour | Dom | Month | DowOptional | Descriptor,
)
```

File: services/actions/schedule_tasks.go
Comment (wolfogre) [inline services/actions/schedule_tasks.go]: `60 * time.Minute` ?
```go
if next.Sub(now) <= 60 {
```

Maybe
```suggestion
	now := time.Now().Truncate(time.Minute)
	for _, row := range specs {
		schedule, err := cron. ParseStandard(row.Spec)
		if err != nil {
			log.Error("ParseSpec: %v", err)
			continue
		}

		if schedule.Next(now.Add(-1)).Equal(now) {
```

File: services/actions/notifier_helper.go
Comment (appleboy) [reply services/actions/notifier_helper.go]: @lunny Done

File: models/actions/schedule.go
Comment (appleboy) [reply models/actions/schedule.go]: Done

File: models/actions/schedule_spec.go
Comment (appleboy) [reply models/actions/schedule_spec.go]: Done

File: services/actions/notifier_helper.go
Comment (appleboy) [reply services/actions/notifier_helper.go]: Done

File: services/actions/schedule_tasks.go
Comment (appleboy) [reply services/actions/schedule_tasks.go]: Done

File: services/actions/schedule_tasks.go
Comment (appleboy) [reply services/actions/schedule_tasks.go]: We also support specific syntax like `@yearly`, `@monthly`, `@weekly`, `@daily`, `@hourly`

File: services/actions/schedule_tasks.go
Comment (appleboy) [inline services/actions/schedule_tasks.go]: use default page size.

File: models/actions/schedule.go
Comment (lunny) [inline models/actions/schedule.go]: Move these to line 48 is better.

File: services/actions/schedule_tasks.go
Comment (appleboy) [reply services/actions/schedule_tasks.go]: Done [0b714b6](https://github.com/go-gitea/gitea/pull/22751/commits/0b714b69ff1f366159e2b6238b5496f51341e7da)

File: models/actions/schedule.go
Comment (appleboy) [reply models/actions/schedule.go]: Done 7671a3124

File: services/actions/schedule_tasks.go
Comment (lunny) [reply services/actions/schedule_tasks.go]: Why should we fetch all before checking them?

Comment (lunny) [reply services/actions/schedule_tasks.go]: Or maybe we could get all ids into memory and then fetch by pagination.

File: services/cron/tasks_actions.go
Comment (appleboy) [reply services/cron/tasks_actions.go]: Done

**Comment by lunny** [comment]
I think this could be refactored after https://gitea.com/gitea/act/pulls/29 merged.

File: modules/actions/workflows.go
Comment (lunny) [inline modules/actions/workflows.go]: Could we also use `OnSchedule` directly?

File: services/actions/schedule_tasks.go
Comment (lunny) [inline services/actions/schedule_tasks.go]:
```suggestion
		if len(specs) < pageSize {
```

File: modules/actions/workflows.go
Comment (appleboy) [reply modules/actions/workflows.go]: Already remove these changes.

**Comment by techknowlogick** [comment]
I've just resolved the conflicts

**Review by silverwind — APPROVED** [review]

File: services/actions/schedule_tasks.go
Comment (lunny) [inline services/actions/schedule_tasks.go]: Looks like `opts` never be used.

Comment (appleboy) [reply services/actions/schedule_tasks.go]: resolved in c7f89f9

**Comment by silverwind** [comment]
This does contain zero tests. Would it be easy to add some? For new features, there is no real excuse to not have at least basic tests.

**Review by lunny — APPROVED** [review]
It's better to have some tests.

**Comment by appleboy** [comment]
@lunny I will rewrite the initialized cron task flow. Don't merge this PR.

**Comment by appleboy** [comment]
Redesign the Cron Job process by adding two fields, prev and next, to the Workflow process to record the execution time. This way, load cron tasks that are less than the current time every minute, avoiding reading the entire table of data. After executing the cron task each time, the system will calculate the next execution time and save it to tables.

cc @lunny @wolfogre @silverwind

**Review by silverwind — APPROVED** [review]

**Comment by silverwind** [comment]
> support specific syntax like @yearly, @monthly, @weekly, @daily, @hourly

FWIW, GitHub Actions decisively does not support this syntax, so by supporting this, we make it harder for people to migrate Gitea to GitHub actions, if this is any concern.

**Comment by silverwind** [comment]
Should this PR contain a docs update?

**Comment by delvh** [comment]
Isn't it the other way around?
If we support it as well while they don't, doesn't it make it harder to migrate away from Gitea to GitHub?
I rather see it as a bonus.
(However, I'n mot sure about the capitalization. The GitHub docs mention it in lowercase, and I would've expected it to be in lowercase as well.)

File: services/actions/notifier_helper.go
Comment (lunny) [inline services/actions/notifier_helper.go]: We need to know which repo and which commit, otherwise this log is useless.

Comment (wxiaoguang) [reply services/actions/notifier_helper.go]: But it's only a trace log for developing. In this case the commit is ignored intentionally, no need to tell end users or site admin. Putting repo into it or not doesn't have much difference (only slightly better ...)

**Comment by lunny** [comment]
> > Isn't it the other way around?
> 
> You misread. I meant migrate from Gitea to GitHub will be harder.
> 
> Regarding docs, why don't we start a page called "Workflow Syntax" [here](https://github.com/go-gitea/gitea/tree/main/docs/content/doc/actions), similar to [this](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions)?
> 
> Edit: retracting review, we need docs before this can land, especially because the syntax differs from GitHub's.

Sure, document PRs are welcome!

**Comment by silverwind** [comment]
> Sure, document PRs are welcome!

I feel like docs for a feature should be part of the feature PR. But I see we are really lacking in terms of actions docs. Still it would be nice to see a syntax document, even if most sections are WIP.

**Review by lunny — APPROVED** [review]

**Review by wxiaoguang — COMMENTED** [review]
Overall LGTM.

The only question in my mind is:

If there is a cron task every minute, there is no active runner. After one hour, will there be 60 tasks in database? Then if a runner starts working, will the runner run these 60 outdated tasks?

**Comment by wxiaoguang** [comment]
> Overall LGTM.
> 
> The only question in my mind is:
> 
> If there is a cron task every minute, there is no active runner. After one hour, will there be 60 tasks in database? Then if a runner starts working, will the runner run these 60 outdated tasks?

@lunny do you have answer to this question?

**Comment by lunny** [comment]
> > Overall LGTM.
> > The only question in my mind is:
> > If there is a cron task every minute, there is no active runner. After one hour, will there be 60 tasks in database? Then if a runner starts working, will the runner run these 60 outdated tasks?
> 
> @lunny do you have answer to this question?

Maybe those tasks should be canceled if it's not running.

**Comment by wxiaoguang** [comment]
> Maybe those tasks should be canceled if it's not running.

Has this PR implemented this? If no, such behavior would cause problems for end users.

**Comment by lunny** [comment]
blocked by #24769

**Comment by silverwind** [comment]
I wouldn't say concurrency groups are strictly blocking this, but some "default"  safeguard to cron tasks would be nice, e.g. something like an implicit "concurrency group".

**Comment by cryptolukas** [comment]
Jenkins have a concept for allowing and denying concurrent builds. Could be adapted.

**Comment by nephatrine** [comment]
While not having concurrency groups might cause hassle for users who decide to use this feature, it seems much more of a hassle to not have any cron schedule tasks at all. There are many uses of this where the concurrency groups wouldn't really even come into play such as a job that runs nightly or weekly - where the job timeout will ensure there aren't multiple pending anyways.

Personally, I'd rather have this feature in 1.20 and take care using it rather than wait for concurrency groups to be supported. This is literally the only thing keeping me on drone instead of moving to gitea's act_runner wholeheartedly.

**Comment by delvh** [comment]
@nephatrine While I feel absolutely the same way about it, 1.20 is out of the picture. We are already in feature freeze, meaning no new features will be accepted for it anymore.

**Comment by BrunoBernardino** [comment]
Sorry for the noise, but I'm so ready for this to be merged to move in a couple of final repos from GitHub into Gitea! :sweat_smile: 

Thank you so much for the hard work here @appleboy !

**Comment by puni9869** [comment]
Could you add some unit test cases.

**Comment by denyskon** [comment]
@appleboy The merged auto-cancellation is only for push events, not for cron, so it won't help us here - or am I wrong?

**Comment by douglasparker** [comment]
Will this make it into gitea soon, or should I look into woodpecker ci or similar?

**Comment by silverwind** [comment]
Happy to rubber-stamp this once the merge conflicts and CI failures are fixed.

**Review by silverwind — APPROVED** [review]

File: models/actions/schedule_list.go
Comment (wxiaoguang) [inline models/actions/schedule_list.go]: It duplicates `ListOptions.ListAll`

**Comment by denyskon** [comment]
Could somebody clarify the state of this PR? IIRC the merged auto-cancellation of jobs only applies to push events. The blocker for this PR was that a wrongly configured cron task could launch many concurrent jobs. So is this PR still blocked, do we not consider this an issue anymore or is there another solution?

File: services/actions/schedule_tasks.go
Comment (wxiaoguang) [inline services/actions/schedule_tasks.go]: continue on err?

File: models/actions/schedule_spec_list.go
Comment (wxiaoguang) [inline models/actions/schedule_spec_list.go]: It seems there is only on call to `FindSpecs` and it doesn't need "count"

File: models/actions/schedule_list.go
Comment (appleboy) [reply models/actions/schedule_list.go]: fixed in 40ed679

File: services/actions/schedule_tasks.go
Comment (appleboy) [reply services/actions/schedule_tasks.go]: fixed in a12d22dc9

File: go.mod
Comment (wxiaoguang) [inline go.mod]: IIRC `gogs/cron` had been removed, now it is back?

Comment (lunny) [reply go.mod]: Yes, this is a block change. It needs to be changed to the new cron package or you could write the parser yourself. @appleboy

**Comment by lunny** [comment]
Sorry, I made a mistake. Please follow #26655

**Pull request closed without merging**
